Move mcumgr functionality as internal Zephyr service

de-nordic opened the pull request:
NOTE: Most of PRs can be squashed, are kept for easier re-editing of the PR contents.

The PR consists of commits:

- affb8f9f66 mgmt/mcumgr: Add CONFIG_IMG_MGMT_FRUGAL_LIST
- beba8296dc manifest: Remove mcumgr from manifest
- fc2f666b29 mgmt/mcumgr/lib: Add Kconfig option MGMT_CBORATTR_FLOAT_SUPPORT
- 2581113690 mgmt/mcumgr/lib: Replace license block
- 053a701a84 mgmt/mcumgr/lib: Fix compliance
- 801ce9d859 mgmt/mcumgr/lib: Documentation fixes
- 35c4c220f2 mgmt/mcumgr/lib: Remove non-Zephyr code
- 49098c1d28 mgmt/mcumgr/lib: Remove log management commands
- 606d3aa251 mgmt/mcumgr/lib: Copy mcumgr library under mgmt subsystem


TODO:
- [x]  Complete testing of MCUMGR.

~~Depends on: https://github.com/zephyrproject-rtos/zephyr/pull/39584~~

Resolves: #36017

de-nordic committed:
mgmt/mcumgr/lib: Copy mcumgr library under mgmt subsystem

The commit copies code of mcumgr, as a lib, under subsys/mgmt/mcumgr.
Only source files relevant for Zephyr have been copied, and changes
to source files have been limited only to path changes, where
required.
Samples have been left out as the only relevant sample, smp_svr,
is already within Zephyr source tree.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
manifest: Remove mcumgr from manifest

The commit removes mcumgr library, as module, from west.yml
manifest, as the mcumgr library source is now kept within Zephyr
source code tree.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

carlescufi commented:
@de-nordic thank you for the PR.
I assume that this is a first step, to later then start modifying the source accordingly in order to get rid of the shim layers that are no longer required now that the code is in-tree. I would also suggest not having a `lib/` folder in the end, but rather the code directly in `subsys/mgmt/mcumgr/` with folders for transports and other logically independent parts.

de-nordic commented:
> 
> 
> @de-nordic thank you for the PR.
> I assume that this is a first step, to later then start modifying the source accordingly in order to get rid of the shim layers that are no longer required now that the code is in-tree. I would also suggest not having a `lib/` folder in the end, but rather the code directly in `subsys/mgmt/mcumgr/` with folders for transports and other logically independent parts.

Yes. This is first step. I want to bring code as it is, of course with filled in missing licenses, compliance and style fixes here and there, so that it would past tests.

Then plan is to change structure of `subsys/mgmg/mcumgr' into something like: `commands`, `protocol`, `transports` and move files around; then specialize and reorganize the code to work efficiently with Zephry and so on.

I want to keep the thing moving and I have decided to make changes in steps;  as I have more priority task to complete and some of them collide, regarding the mcumgr changes, I need to structure work in a way that lets me keep the changes going in gradually while working on other items.

de-nordic committed:
mgmt/mcumgr/lib: Fix compliance

Commit fixes compliance issues.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr/lib: Remove log management commands

Not implemented.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr/lib: Remove non-Zephyr code

The commit removes MYNEWT sepcific code and code that has been
conditionally compiled with "ifndef _ZEPHYR_".

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr/lib: Replace license block

The commit changes licensing to SPDX tags.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr: Add CONFIG_IMG_MGMT_FRUGAL_LIST

The commit adds Kconfig options that enables "frugal status list";
when enabled, the status list will be sent with zero, empty and false
values omitted, slightly reducing number of bytes sent as response
from device.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

carlescufi commented:
@rerickson1 and @lairdjm please review

rerickson1 commented:
@ahedin25

rerickson1 commented on subsys/mgmt/mcumgr/lib/README.md:
```diff
@@ -0,0 +1,137 @@
+For tips on using mcumgr with your particular OS, see the appropriate file from
+the list below:
+
+* [Mynewt](README-mynewt.md)
```
Not needed

de-nordic replied:
Removed

rerickson1 commented on subsys/mgmt/mcumgr/lib/README.md:
```diff
@@ -0,0 +1,137 @@
+
+## Getting started
+
+For tips on using mcumgr with your particular OS, see the appropriate file from
+the list below:
```
probably should be re-worded because Zephyr is our only OS.

de-nordic replied:
Removed

rerickson1 commented on subsys/mgmt/mcumgr/lib/README.md:
```diff
@@ -0,0 +1,137 @@
+- [cmd](cmd): Built-in command handlers for the core mcumgr commands.
+- [ext](ext): Third-party libraries that mcumgr depends on.
+- [mgmt](mgmt): Code implementing the `mgmt` layer of mcumgr.
+- [samples](samples): Sample applications utilizing mcumgr.
```
doesn't exist anymore

de-nordic replied:
Fixed

rerickson1 commented on subsys/mgmt/mcumgr/lib/README-zephyr.md:
```diff
@@ -11,5 +11,4 @@ are described below:
 | `CONFIG_MCUMGR` | Enable the mcumgr management library. | n |
 | `CONFIG_MCUMGR_CMD_FS_MGMT` | Enable mcumgr handlers for file management | n |
 | `CONFIG_MCUMGR_CMD_IMG_MGMT` | Enable mcumgr handlers for image management | n |
-| `CONFIG_MCUMGR_CMD_LOG_MGMT` | Enable mcumgr handlers for log management | n |
```
Need to squash this commit yet or change the commit description?
Leaving this commit may make the changes more readable.

de-nordic replied:
Sorry, you need to be more verbose because I did not get that.

rerickson1 replied:
The Commit description for the commit containing this change says it will be squashed. You should squash this commit or remove that message in the commit description.
I would be in favor of leaving this commit to have better tracking of the changes.

de-nordic replied:
Removed information on squashing from commit messages.

lairdjm commented on subsys/mgmt/mcumgr/lib/README-zephyr.md:
```diff
@@ -0,0 +1,14 @@
+| `CONFIG_MCUMGR` | Enable the mcumgr management library. | n |
+| `CONFIG_MCUMGR_CMD_FS_MGMT` | Enable mcumgr handlers for file management | n |
+| `CONFIG_MCUMGR_CMD_IMG_MGMT` | Enable mcumgr handlers for image management | n |
+| `CONFIG_MCUMGR_CMD_OS_MGMT` | Enable mcumgr handlers for OS management | n |
```
Missing CONFIG_MCUMGR_GRP_ZEPHYR_BASIC

de-nordic replied:
Done

lairdjm commented on subsys/mgmt/mcumgr/lib/protocol.md:
```diff
@@ -0,0 +1,256 @@
+
+## Transports
+
+### Mcumgr/Newtmgr SMP Client Over Serial
+
+`mcumgr` or `newtmgr` can be used over TTY serial with the following parameters
```
Does newtmgr still work with zephyr?

de-nordic replied:
Yes, mcumgr is just rename for newtmgr

de-nordic committed:
mgmt/mcumgr/lib: Documentation fixes

References to Mynewt have been removed and README for Zephyr has been
merged directly to the README.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr/lib: Add Kconfig option MGMT_CBORATTR_FLOAT_SUPPORT

The option enables float numbers processing within CBOR attributes.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic commented on subsys/mgmt/mcumgr/lib/README.md:
```diff
@@ -0,0 +1,137 @@
+
+To use mcumgr's image management support, your device must be running version
+1.1.0 or later of the [MCUboot boot
+loader](https://github.com/runtimeco/mcuboot).  The other mcumgr features do
```
Done

de-nordic commented on subsys/mgmt/mcumgr/lib/README.md:
```diff
@@ -0,0 +1,137 @@
+* Command handlers:
+    * Image management (`img_mgmt`)
+    * File system management (`fs_mgmt`)
+    * Log management (`log_mgmt`)
```
Done

de-nordic commented on subsys/mgmt/mcumgr/lib/README.md:
```diff
@@ -0,0 +1,137 @@
+
+Developers welcome!
+
+* Our Slack channel: https://mynewt.slack.com/messages/C7Y3K0C2J
```
Done

de-nordic commented on subsys/mgmt/mcumgr/lib/cborattr/src/cborattr.c:
```diff
@@ -0,0 +1,404 @@
+			lptr = &arr->arr.uintegers.store[off];
+			err |= cbor_value_get_uint64(&elem, lptr);
+			break;
+#if FLOAT_SUPPORT
```
Done

de-nordic commented on subsys/mgmt/mcumgr/lib/cborattr/test/src/test_cborattr.h:
```diff
@@ -0,0 +1,46 @@
+TEST_CASE_DECL(test_cborattr_decode_string_array);
+TEST_CASE_DECL(test_cborattr_decode_object_array);
+TEST_CASE_DECL(test_cborattr_decode_unnamed_array);
+TEST_CASE_DECL(test_cborattr_decode_substring_key);
```
Yes we will have to do that.

de-nordic commented on subsys/mgmt/mcumgr/lib/mgmt/include/mgmt/mgmt.h:
```diff
@@ -0,0 +1,429 @@
+#endif
+
+/* MTU for newtmgr responses */
+#define MGMT_MAX_MTU		1024
```
Unfortunately right now this thing is not used anywhere.

de-nordic commented:
@lairdjm Can you ping me on discord so we could have brief talk on further plans for the mcumgr?

rerickson1 commented on subsys/mgmt/mcumgr/lib/README.md:
```diff
@@ -8,22 +8,28 @@ components.  In addition, mcumgr provides definitions and handlers for some
+| `CONFIG_MCUMGR_CMD_FS_MGMT` | Enable mcumgr handlers for file management | n |
+| `CONFIG_MCUMGR_CMD_IMG_MGMT` | Enable mcumgr handlers for image management | n |
+| `CONFIG_MCUMGR_CMD_OS_MGMT` | Enable mcumgr handlers for OS management | n |
+| `CONFIG_MCUMGR_CMD_STAT_MGMTi` | Enable mcumgr handlers for statistics | n |
```
typo `_MGMTi`

de-nordic replied:
Fixed

rerickson1 commented on subsys/mgmt/mcumgr/lib/README.md:
````diff
@@ -0,0 +1,141 @@
+
+```
++----------+----------+----------+----------+
+| img_mgmt |  fs_mgmt | log_mgmt |  os_mgmt |
````
log_mgmt no longer applies

de-nordic replied:
Done

de-nordic commented on subsys/mgmt/mcumgr/lib/cmd/img_mgmt/include/img_mgmt/img_mgmt.h:
```diff
@@ -0,0 +1,269 @@
+ */
+#define IMG_MGMT_STATE_F_PENDING	0x01
+#define IMG_MGMT_STATE_F_CONFIRMED	0x02
+#define IMG_MGMT_STATE_F_ACTIVE		0x04
+#define IMG_MGMT_STATE_F_PERMANENT	0x08
+
+#define IMG_MGMT_VER_MAX_STR_LEN	25  /* 255.255.65535.4294967295\0 */
+
+/*
+ * Swap Types for image management state machine
+ */
+#define IMG_MGMT_SWAP_TYPE_NONE		0
+#define IMG_MGMT_SWAP_TYPE_TEST		1
+#define IMG_MGMT_SWAP_TYPE_PERM		2
+#define IMG_MGMT_SWAP_TYPE_REVERT	3
+
```
Fixed
